json/api/compose_data.go

```diff
@@ -0,0 +1,232 @@
+/*
+* Copyright(c) 2012-2021 Intel Corporation
+* SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+package api
+
+import (
+	"fmt"
+	"os"
+
+	"example.com/json/ioctl"
+)
+
+/** packages of JSON marshalable structs with packed CAS info for easy use in RESTful API */
+
+type Cache_with_cores struct {
+	// default value of ptr is nil
+	Cache_info *ioctl.Kcas_cache_info  `json:"Cache"`
+	Cores_info []*ioctl.Kcas_core_info `json:"Cores"`
+}
+
+type Stats_pkg struct {
+	Cache_info *ioctl.Kcas_cache_info `json:"Cache info"`
+	Core_info  *ioctl.Kcas_core_info  `json:"Core info"`
+	Io_class   *ioctl.Kcas_io_class   `json:"IO class"`
+	Stats      *ioctl.Kcas_get_stats  `json:"Statistics"`
+}
+
+type Cache_list_pkg []Cache_with_cores
+
+/** Retriving caches and their cores info from CAS */
+
+func List_caches() (Cache_list_pkg, error) {
+	var cache_list_pkg Cache_list_pkg
+
+	/** retrive file descriptor of exported object cas_ctrl
+	and close with defer after completing all tasks */
+	cas_ctrl, fd_err := os.Open(ioctl.Cas_ctrl_path)
+	defer cas_ctrl.Close()
```

Comment:
If the `Open` operation fails, there is nothing to be closed.

Reply:
placing defer after early return statement should be fix  for that problem.